server/base/src/test/java/org/apache/accumulo/server/conf/codec/VersionedPropEncryptCodec.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.conf.codec;
+
+import static org.apache.accumulo.server.conf.codec.EncodingOptions.EXPERIMENTAL_CIPHER_ENCODING_1_0;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * EXPERIMENTAL - demonstrates using an alternate encoding scheme. The sample is completely
+ * functional, however, certain elements such as password / key handling may not be suitable for
+ * production. The encoding version is EXPERIMENTAL_CIPHER_ENCODING_1_0.
+ * <p>
+ * This codec uses AES algorithm in GCM mode for encryption to encode the property map that is
+ * stored in the external store.
+ */
+public class VersionedPropEncryptCodec extends VersionedPropCodec {
```

Comment:
Could make the `@Experimental` annotation public and use it here. It is currently only used by `Property` keys but I don't see why we couldn't use it elsewhere in places like here.

Also, I noticed this class is in the Test package. I am not sure that was intentional or not. If you put it in Test because its experimental, maybe using the annotation would make you feel better about putting it with the other codecs.

Reply:
I put in test until we discuss if it is even a desirable to provide as an option. It was really useful to drive the design because it required additional metadata - if it is acceptable and moved into the code base, there would be small changes required (like a property for codec selection) - those would be worthwhile to do, but I deferred until they would actually be used.  

The code should also receive additional scrutiny to check that I'm not just providing an illusion of increased confidentially because of a coding or misapplication of the encryption API.

In test, it seemed clear enough to show intent and how additional codecs could be developed is desired.

Reply:
Would use of the `@Experimental` annotation alleviate your concerns about putting it in src?

Reply:
I can go either way.  The one sticking point for me is how to treat the secret - maybe you have some guidance from your other work with encryption?  It may be external to these changes, but in the scope of making it an option (even if experimental) I think we would need to provide something.

Reply:
The `AESCryptoService` uses GCM similar to how you are using it but there are some details that differ. I don't know the imp super well but some differences that I noticed between yours and `AESCryptoService`:
- It uses `SecureRandom` to generate the key
- IV gets stored in memory so it can be incremented
- There is some Key wrapping/unwrapping
- Key gets loaded from a URI